### PR TITLE
Implement -o flag for a few functions

### LINKED ISSUE #897: add -o flag in sct_smooth_spinalcord

### LINKED ISSUE #1802: Implement -o flag for a few functions
### Description

In addition to the already-existing `-ofolder` for putting the outputs there, it would be useful to also have a `-o` flag to specify output file (not folder). Functions where we could implement that are:
- sct_propseg
- sct_deepseg_sc
- sct_deepseg_gm

### PR DESCRIPTION
**Pull request by lrouhier**
#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've assigned a reviewer

#### PR contents

- [x] I've consulted [SCT's internal developer documentation](https://github.com/neuropoly/spinalcordtoolbox/wiki) to ensure my contribution is in line with any relevant design decisions
- [x] I've added [relevant tests](https://github.com/neuropoly/spinalcordtoolbox/wiki/Programming%3A-Tests) for my contribution
- [x] I've updated the [relevant documentation](https://github.com/neuropoly/spinalcordtoolbox/wiki/Programming%3A-Documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description
This PR aims at adding the `-o` flag on a few function in sct:
- [x] sct_deepseg_sc
- [x] sct_detect_pmj
- [x] sct_propseg
- [x] sct_smooth_spinalcord

## Linked issues
fix #1802
fix #897

### CODE REVIEW

File: spinalcordtoolbox/scripts/sct_smooth_spinalcord.py
Comment (jcohenadad) [inline spinalcordtoolbox/scripts/sct_smooth_spinalcord.py:143]: please make sure to not change the default behaviour-- in this case if input is NII the output will be NIIGZ (whereas the current default is NII)

File: spinalcordtoolbox/scripts/sct_deepseg_sc.py
Comment (jcohenadad) [inline spinalcordtoolbox/scripts/sct_deepseg_sc.py:175]: there is probably a more elegant way than launching the same API twice (extract_fname), maybe @joshuacwnewton you have some ideas?

Comment (lrouhier) [reply spinalcordtoolbox/scripts/sct_deepseg_sc.py:175]: We can run it only once and to get the all the output and then use that.

File: spinalcordtoolbox/scripts/sct_smooth_spinalcord.py
Comment (lrouhier) [reply spinalcordtoolbox/scripts/sct_smooth_spinalcord.py:143]: Ah ! Sorry I'll change it.

**Comment by jcohenadad** [comment]
it might be good to create tests for these, no?  e.g. test if -o works as expected-- and check maybe the default behaviour?

File: unit_testing/cli/test_cli_sct_detect_pmj.py
Comment (joshuacwnewton) [inline unit_testing/cli/test_cli_sct_detect_pmj.py:18]: Same comments re: subprocess and `tmp_path`.

File: unit_testing/cli/test_cli_sct_propseg.py
Comment (joshuacwnewton) [inline unit_testing/cli/test_cli_sct_propseg.py:19]: Same comments re: subprocess and `tmp_path`.

File: unit_testing/cli/test_cli_sct_smooth_spinalcord.py
Comment (joshuacwnewton) [inline unit_testing/cli/test_cli_sct_smooth_spinalcord.py:19]: Same comments re: subprocess and `tmp_path`.

**Review by joshuacwnewton — CHANGES_REQUESTED** [review]
Thanks for your ongoing efforts to tackle all of these older issues, @lrouhier!

File: spinalcordtoolbox/scripts/sct_smooth_spinalcord.py
Comment (joshuacwnewton) [inline spinalcordtoolbox/scripts/sct_smooth_spinalcord.py:257]: Here's the relevant change for one of my review comments.

File: unit_testing/cli/test_cli_sct_deepseg_sc.py
Comment (lrouhier) [reply unit_testing/cli/test_cli_sct_deepseg_sc.py:23]: I will look into the tmp_path fixture. 
However for the sct_function.main(args), I do have an issue. As you can see [here](https://github.com/neuropoly/spinalcordtoolbox/blob/dc97d42730d7e603f83e54c68f242ad2bef7189b/spinalcordtoolbox/scripts/sct_deepseg_sc.py#L130), the sct_deepseg.main() function does not take any argument. Should I change that for the sake of the test ? (this is not the only one)

File: unit_testing/cli/test_cli_sct_propseg.py
Comment (lrouhier) [reply unit_testing/cli/test_cli_sct_propseg.py:19]: About subprocess: [Here](https://github.com/neuropoly/spinalcordtoolbox/blob/dc97d42730d7e603f83e54c68f242ad2bef7189b/spinalcordtoolbox/scripts/sct_propseg.py#L657-L676) sct_propseg needs the output of the parser as it is not called inside the 'main' function.

File: unit_testing/cli/test_cli_sct_smooth_spinalcord.py
Comment (lrouhier) [reply unit_testing/cli/test_cli_sct_smooth_spinalcord.py:19]: [Here](https://github.com/neuropoly/spinalcordtoolbox/blob/dc97d42730d7e603f83e54c68f242ad2bef7189b/spinalcordtoolbox/scripts/sct_smooth_spinalcord.py#L120-L127), it does not take into consideration if args are passed to main (only if they are passed through the sys)

File: spinalcordtoolbox/scripts/sct_smooth_spinalcord.py
Comment (lrouhier) [reply spinalcordtoolbox/scripts/sct_smooth_spinalcord.py:246]: They are not equal, actually, the change that you showed in your [comment](https://github.com/neuropoly/spinalcordtoolbox/pull/3081#discussion_r537536763) was necessary due to a mistake. Indeed it gave you the command to open the file but there was no extension. So now it displays the full name.

File: unit_testing/cli/test_cli_sct_deepseg_sc.py
Comment (joshuacwnewton) [reply unit_testing/cli/test_cli_sct_deepseg_sc.py]: Ahh, good point. It looks like some `main()` functions have been edited to take args but not all. I'll make an issue for this. :)

File: unit_testing/cli/test_cli_sct_propseg.py
Comment (joshuacwnewton) [reply unit_testing/cli/test_cli_sct_propseg.py]: Interesting... thanks for pointing that out! Yes, this is definitely inconsistent between scripts.

File: unit_testing/cli/test_cli_sct_detect_pmj.py
Comment (lrouhier) [reply unit_testing/cli/test_cli_sct_detect_pmj.py]: Tmp_path fixed. [main ](https://github.com/neuropoly/spinalcordtoolbox/blob/dc97d42730d7e603f83e54c68f242ad2bef7189b/spinalcordtoolbox/scripts/sct_detect_pmj.py#L260) does not take args.

File: spinalcordtoolbox/scripts/sct_smooth_spinalcord.py
Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_smooth_spinalcord.py:246]: Ahhh, I see. Thank you for clarifying.

File: unit_testing/cli/test_cli_sct_smooth_spinalcord.py
Comment (lrouhier) [reply unit_testing/cli/test_cli_sct_smooth_spinalcord.py]: tmp_path fixed

File: spinalcordtoolbox/scripts/sct_detect_pmj.py
Comment (lrouhier) [reply spinalcordtoolbox/scripts/sct_detect_pmj.py:291]: Right. I guess we could directly go with fname_o = path_result+arguments.o

Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_detect_pmj.py:291]: Hmm... the `DetectPMJ` class is written in a way where that wouldn't work:

https://github.com/neuropoly/spinalcordtoolbox/blob/1b220017c7f48613f3168a33a9dbd7d72eab4bdf/spinalcordtoolbox/scripts/sct_detect_pmj.py#L159-L160

So, `DetectPMJ`'s `fname_out` attribute has to be in the form filename+ext without the path... which makes things a little weird.

Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_detect_pmj.py:291]: Maybe this is too much effort for too little gain. :sweat_smile:

File: unit_testing/cli/test_cli_sct_deepseg_sc.py
Comment (joshuacwnewton) [reply unit_testing/cli/test_cli_sct_deepseg_sc.py]: I've opened #3085. We could either wait for discussion there, or just leave the tests as subprocess for now and add a note. Something like:

```python
# TODO: Change to main(args) call once Issue #3085 has been resolved
subprocess.check_output(command, shell=True)
```

Comment (joshuacwnewton) [reply unit_testing/cli/test_cli_sct_deepseg_sc.py]: #3091 has been opened to fix #3085. 

Once #3091 gets merged, then you will be able to use `sct_deepseg_sc.main` (or the `main` of any SCT script!)

File: unit_testing/cli/test_cli_sct_deepseg_sc.py
Comment (jcohenadad) [reply unit_testing/cli/test_cli_sct_deepseg_sc.py]: > Perhaps in a new issue?

sounds good to me

Comment (joshuacwnewton) [reply unit_testing/cli/test_cli_sct_deepseg_sc.py]: Done in #3093.

**Comment by jcohenadad** [comment]
@lrouhier @joshuacwnewton what is the status of this PR, can it be updated and merged?

**Comment by joshuacwnewton** [comment]
> @lrouhier @joshuacwnewton what is the status of this PR, can it be updated and merged?

I think it's very close to done, and was just waiting on #3091 to allow relevant sct scripts to be tested using `main()` (https://github.com/neuropoly/spinalcordtoolbox/pull/3081#discussion_r541060418). Since #3091 now has approval, I can merge it, and this one can be finished up.

**Comment by lrouhier** [comment]
@joshuacwnewton, I think this is fairly ready now. I'm just waiting on the test but everything should run smoothly. If you could review the PR it would be great!

**Comment by lrouhier** [comment]
The sct_propseg test is not using the tmp_path fixture because there was an error with the `mv` function from  spinalcordtoolbox.utils.fs see [this log](https://travis-ci.com/github/neuropoly/spinalcordtoolbox/jobs/465807495).

**Comment by joshuacwnewton** [comment]
> The sct_propseg test is not using the tmp_path fixture because there was an error with the `mv` function from spinalcordtoolbox.utils.fs see [this log](https://travis-ci.com/github/neuropoly/spinalcordtoolbox/jobs/465807495).

Hmm... I don't think this is necessarily a problem with `tmp_path`. Instead, I think this means we [should be using `shutil.move`](https://stackoverflow.com/a/43967659) instead of `os.rename`.

I'm going to try it out in a quick PR and if all goes well then `tmp_path` should be usable here too.

File: unit_testing/cli/test_cli_sct_deepseg_sc.py
Comment (joshuacwnewton) [inline unit_testing/cli/test_cli_sct_deepseg_sc.py:23]: `tmp_path` returns a `pathlib.Path` object, which [has some useful properties](https://realpython.com/python-pathlib/). You can use `/` to combine paths instead of `os.path.join`! And, when combined with f-strings (`f""`), you don't need any of the `str()` conversion, so it can be written in a short way:

```suggestion
    tmp_filepath = tmp_path/'test_seg.nii.gz'
    command = f"-i sct_testing_data/t2/t2.nii.gz -c t2 -o {tmp_filepath}"
    sct_deepseg_sc.main(command.split())
    assert os.path.isfile(tmp_filepath)
```

Comment (joshuacwnewton) [reply unit_testing/cli/test_cli_sct_deepseg_sc.py:23]: This same refactor could apply to the other tests, too. :)

File: unit_testing/cli/test_cli_sct_deepseg_sc.py
Comment (joshuacwnewton) [reply unit_testing/cli/test_cli_sct_deepseg_sc.py:6]: Ditto for the other tests!

**Review by joshuacwnewton — CHANGES_REQUESTED** [review]
I have a few small style nitpicks.

File: unit_testing/cli/test_cli_sct_detect_pmj.py
Comment (kousu) [inline unit_testing/cli/test_cli_sct_detect_pmj.py:21]: Heyo, how's it gong? I'm parachuting in here from https://github.com/neuropoly/spinalcordtoolbox/pull/3146 to understand context. I noticed here that this split() isn't really safe. It's not quite as prone to problems as `os.system()` but it's still vulnerable to mistakes. If some system sometime happens to have spaces in its `$TMPDIR` then this code is going to have a bad time.

The thing that people usually do with `subprocess.run()` is to pass ARGV explicitly, and I think that's a better option here:

```suggestion
    argv = ['-i', 'sct_testing_data/template/template//PAM50_small_t2.nii.gz', '-igt','sct_testing_data/template/template/PAM50_small_t2_pmj_manual.nii.gz', '-c', 't2', '-o', os.path.join(str(tmp_path), 'test_pmj.nii.gz')]
    sct_detect_pmj.main(argv)
```

I see you already removed this tmp_path from the other test, and that sidesteps the issue: the string is static so it can't have extra spaces slip in to mess it up. But it would be more reliable to just make it a list from the start.

File: spinalcordtoolbox/scripts/sct_deepseg_sc.py
Comment (kousu) [inline spinalcordtoolbox/scripts/sct_deepseg_sc.py:98]: All the other options use single quotes

```suggestion
        '-o',
```

File: spinalcordtoolbox/scripts/sct_detect_pmj.py
Comment (kousu) [inline spinalcordtoolbox/scripts/sct_detect_pmj.py:76]:
```suggestion
        '-o',
```

File: spinalcordtoolbox/scripts/sct_deepseg_sc.py
Comment (kousu) [inline spinalcordtoolbox/scripts/sct_deepseg_sc.py:102]: oh, except not this one line? but the other scripts all do use single quotes.

```suggestion
        '-r',
```

File: spinalcordtoolbox/scripts/sct_propseg.py
Comment (kousu) [inline spinalcordtoolbox/scripts/sct_propseg.py:214]:
```suggestion
        '-o',
```

File: spinalcordtoolbox/scripts/sct_propseg.py
Comment (kousu) [inline spinalcordtoolbox/scripts/sct_propseg.py:462]: This is the cause of the error in https://github.com/neuropoly/spinalcordtoolbox/pull/3081#issuecomment-752034463. It needs to be aware of `-ofolder` too.

**Probably** this PR should **remove** `-ofolder` entirely since `-o` covers the same purpose. You should replace the above `folder_out` section with:

```
folder_out = os.path.dirname(arguments.o)
```

If you need to keep both arguments -- which I really recommend against -- then you need to make sure they are consistent with each other. That means doing something like:

```
# quick draft, don't commit as is, please test it out
folder_out, fname_out = os.path.split(arguments.o)
if arguments.ofolder is not None and arguments.ofolder != folder_out:
    raise ValueError(f"Output file {arguments.o} must be in output folder {arguments.ofolder}") 
```

What's `-ofolder` for anyway? Does `propseg` output extra metadata to that folder?

Comment (kousu) [reply spinalcordtoolbox/scripts/sct_propseg.py:462]: By the time it gets to

https://github.com/neuropoly/spinalcordtoolbox/blob/552b908c63234c23cf65e9e520ec3c88a55bcec8/spinalcordtoolbox/scripts/sct_propseg.py#L635-L640

you must have

```
assert not os.path.dirname(fname_out)
```

That's what that section of code is implicitly assuming. You can either fix it there, or fix it above. Adding that assertion is probably a good idea, at least for the tests.

Comment (kousu) [reply spinalcordtoolbox/scripts/sct_propseg.py:462]: Ou.... j'ai pas encore tout lire. Il y ..trois? fichiers? `f"{F}_rescaled.nii.gz"` et `f"{F}_rescaled_centerline.nii.gz"` et ça veux redire vers `f"{F}.nii.gz"` et `f"{F}_centerline.nii.gz"?

Est-ce qu'on c'est possible de juste demander de `isct_propseg` de faire le bon chose en prèmiere?

Comment (jcohenadad) [reply spinalcordtoolbox/scripts/sct_propseg.py:458]: > What's -ofolder for anyway? Does propseg output extra metadata to that folder?

propseg can output more than one stuff, e.g. `-CSF`, ` -centerline-coord`. Historically, the `-ofolder` was created when we were using `sct_testing`: this script included a function that used the flag `-ofolder`, when it existed in some scripts. Since then, we completely refactored the testing suite and we don't use -ofolder anymore, so I agree we could get rid of -ofolder as it is confusing.

But we need to keep in mind the multiple (possible) outputs. However, since these are not widely used, we can keep -o to specify the output of the cord segmentation (primary interest for sct_propseg) and have the other outputs with fix naming (very few people use the other outputs anyway).

File: unit_testing/cli/test_cli_sct_deepseg_sc.py
Comment (lrouhier) [reply unit_testing/cli/test_cli_sct_deepseg_sc.py]: Done

File: spinalcordtoolbox/scripts/sct_propseg.py
Comment (lrouhier) [reply spinalcordtoolbox/scripts/sct_propseg.py:449]: I tried something to remove this -ofolder issue, tell me what you think! (bab7bff)

File: unit_testing/cli/test_cli_sct_deepseg_sc.py
Comment (lrouhier) [reply unit_testing/cli/test_cli_sct_deepseg_sc.py]: I went with argv so no need for formatted string ;)

File: spinalcordtoolbox/scripts/sct_propseg.py
Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_propseg.py:449]: If the -ofolder issue is fixed, `tmp_path` should be usable again, so could you add `tmp_path` back to the `sct_propseg` test?  Would help to test this change. :slightly_smiling_face:

Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_propseg.py:449]: To clarify what @kousu was saying, the error occurs with this line:

https://github.com/neuropoly/spinalcordtoolbox/blob/fd00c800467aa194fcbae073eb10bc589d772920/spinalcordtoolbox/scripts/sct_propseg.py#L639-L640

When called in CI, the two values are:

Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_propseg.py:449]: To clarify what @kousu was saying, the error occurs with this line:

https://github.com/neuropoly/spinalcordtoolbox/blob/fd00c800467aa194fcbae073eb10bc589d772920/spinalcordtoolbox/scripts/sct_propseg.py#L639-L640

When this command is called in the test, the two values are this: [(CI link)](https://travis-ci.com/github/neuropoly/spinalcordtoolbox/jobs/465807495#L2096-L2105)

```
src = './t2_rescaled_seg.nii.gz'
dst = '/tmp/pytest-of-travis/pytest-0/popen-gw0/test_sct_propseg_o_flag0/test_seg.nii.gz'
```

The problem here is that `src` is not using `tmp_path`. (Every output file should be created in the `tmp_path` directory.)

To fix this, you need to make sure that `folder_output` uses `tmp_path`.

https://github.com/neuropoly/spinalcordtoolbox/blob/3df9aeffcade01e30befd9b6074fc77663d051e5/spinalcordtoolbox/scripts/sct_propseg.py#L438

So, you might use [`os.path.dirname`](https://docs.python.org/3/library/os.path.html#os.path.dirname) get the parent directory from the `-o` argument, so that everything is output to that directory.

```diff
- folder_output = './' 
+ folder_output = os.path.dirname(fname_out)  # Get base directory from output path
```

Comment (lrouhier) [reply spinalcordtoolbox/scripts/sct_propseg.py]: @joshuacwnewton what do you think about f81733d ?

File: unit_testing/cli/test_cli_sct_smooth_spinalcord.py
Comment (joshuacwnewton) [inline unit_testing/cli/test_cli_sct_smooth_spinalcord.py:4]:
```suggestion
```

File: unit_testing/cli/test_cli_sct_propseg.py
Comment (joshuacwnewton) [inline unit_testing/cli/test_cli_sct_propseg.py:4]:
```suggestion
```

File: unit_testing/cli/test_cli_sct_deepseg_sc.py
Comment (joshuacwnewton) [inline unit_testing/cli/test_cli_sct_deepseg_sc.py:5]:
```suggestion
```

**Review by joshuacwnewton — APPROVED** [review]
I have some small nits, but they're very straightforward to fix, so I'm pre-approving.

File: spinalcordtoolbox/scripts/sct_propseg.py
Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_propseg.py]: Looks good to me! :)

**Pull request merged**
